### PR TITLE
Added changing of languages runtime within the menu

### LINKED ISSUE #246: Add option to set in game language
As the title says, add an option (In game menu and xml) to set the in game language. Right now we have to do the following to launch performous in another language:
```
export LANG=nl_NL.UTF-8
export LANGUAGE=nl_NL
```
This is not user friendly :joy:

Referenced from #236 

# TODO:
- [x] Add menu item
- [x] Show only 100% translated languages??
    - [x] It defaults to English when there's no translation found. If objection please mention
- [x] Save it to config.xml
- [x] Load the language at runtime
- [x] ~~Include #242~~

### PR DESCRIPTION
**Pull request by Baklap4**
### What does this PR do?

Adds changing the language while in the game menu.
It also starts the game in the language you're playing in.
The languages are also translated into the language you're viewing performous.

### Closes Issue(s)

Closes #246 

### Motivation

As explained in #246 we could only change the language by setting some environment variables. This was not quite handy. Fixed it so we can do it from within the menu

### Testing Notes

If you want to test this you should install the app. Otherwise you have to create a folder structure (within your build folder) like this:
```
build
|--- lang
|----- nl.mo
|----- nl
|------- LC_MESSAGES
|---------- Performous.mo
```

### CODE REVIEW

File: game/configuration.cc [hunk 1]
```diff
@@ -293,6 +311,16 @@ void writeConfig(bool system) {
 				Game::getSingletonPtr()->window().resetShaders();
 			}
 		}
+		if(name == "game/language") {
+			std::string oldValue = Game::getSingletonPtr()->getCurrentLanguage();
+			std::string newValue = item.getEnumName();
```
Comment (OznOg) [inline game/configuration.cc:316]: same

File: game/i18n.hh [hunk 2]
```diff
@@ -2,26 +2,74 @@
...
+		std::clog << "locale/debug: Checking if language is set in config." << std::endl;
+		if(config["game/language"].i() >= 0) {
+			lang = config["game/language"].getEnumStringValueAtIndex(config["game/language"].i());
+			std::clog << "locale/debug: Language in configuration found: " << lang << std::endl;
+		}
+
+		if (lang.empty()) lang = "en_US.UTF-8";
+		setLanguage(lang);
+	};
+	static bool enabled() {
```
Comment (OznOg) [inline game/i18n.hh:26]: seems useless

File: game/i18n.hh [hunk 3]
```diff
@@ -2,26 +2,74 @@
...
+		return true;
+	};
+
+	void setLanguage(const std::string language) {
```
Comment (OznOg) [inline game/i18n.hh:30]: const& or useless const

File: game/i18n.hh [hunk 4]
```diff
@@ -2,26 +2,74 @@
...
 			std::clog << "locale/warning: Unable to detect locale, will try to fallback to en_US.UTF-8" << std::endl;
 			std::locale::global(gen("en_US.UTF-8"));
 		}
-	};
-	static bool enabled() {
-		return true;
-	};
+	}
+
+	const std::string getLanguage(const std::string language) {
```
Comment (OznOg) [inline game/i18n.hh:51]: no need for const on return value; const & on parameter

File: game/i18n.hh [hunk 5]
```diff
@@ -2,26 +2,74 @@
...
+
+		if (foundLanguage == allLanguages.end())
+			returnLang = "en_US.UTF-8";
+		else 
+			returnLang = *foundLanguage;
+
+		return returnLang;
+	}
 
+	const std::string getCurrentLanguage() { return m_currentLanguage; }
```
Comment (OznOg) [inline game/i18n.hh:70]: same

File: game/i18n.hh [hunk 6]
```diff
@@ -2,26 +2,74 @@
...
+	std::vector<std::string> GetAllLanguages() { return config["game/language"].getAllEnumStringValues(); }
+private:
+	std::string m_currentLanguage;
+	const char *m_package;
```
Comment (OznOg) [inline game/i18n.hh:74]: can it be null? who has ownership of that data? maybe considere using unique_ptr, or reference if not owned

File: game/screen.hh [hunk 7]
```diff
@@ -103,6 +105,9 @@ class Game: public Singleton <Game> {
 	void drawLogo();
 	///global playlist access
 	PlayList& getCurrentPlayList() { return currentPlaylist; }
+	void setLanguage(const std::string language) { m_translationEngine.setLanguage(language); };
+    const std::string getCurrentLanguage() { return m_translationEngine.getCurrentLanguage(); };
```
Comment (OznOg) [inline game/screen.hh:109]: non const on return value

File: game/configuration.cc [hunk 8]
```diff
@@ -165,6 +165,24 @@ namespace {
 	}
 }
 
+std::string ConfigItem::getEnumStringValueAtIndex(unsigned int index) {
+	try {
+		auto item = m_enums.at(index);
+		return item;
```
[the remark this answers is not in the document]
Comment (Baklap4) [reply game/configuration.cc:171]: At throws a index out of range exception better catch it and throw runtime error

File: game/configuration.cc [hunk 9]
```diff
@@ -165,6 +165,24 @@ namespace {
...
+	} catch (const std::out_of_range& e) {
+        std::clog << "configuration/error: Out of Range error." << e.what() << std::endl;
+		throw std::runtime_error("Enum Index not found in enumeration.");
```
[the remark this answers is not in the document]
Comment (Baklap4) [reply game/configuration.cc:174]: Alright will change :)

File: game/configuration.cc [hunk 10]
```diff
@@ -293,6 +311,16 @@ void writeConfig(bool system) {
 				Game::getSingletonPtr()->window().resetShaders();
 			}
 		}
+		if(name == "game/language") {
```
[the remark this answers is not in the document]
Comment (Baklap4) [reply game/configuration.cc:314]: It was late gonna stick with `if (`

File: game/configuration.cc [hunk 11]
```diff
@@ -293,6 +311,16 @@ void writeConfig(bool system) {
...
+			std::string oldValue = Game::getSingletonPtr()->getCurrentLanguage();
```
[the remark this answers is not in the document]
Comment (Baklap4) [reply game/configuration.cc:315]: Will change :+1:

File: game/configuration.cc (hunk 1 above)
Comment (Baklap4) [reply game/configuration.cc:316]: Will change

File: game/i18n.hh (hunk 2 above)
Comment (Baklap4) [reply game/i18n.hh:26]: Not sure but will check it to make it sure :+1:

File: game/i18n.hh (hunk 3 above)
Comment (Baklap4) [reply game/i18n.hh:30]: Yup gonna get changed

File: game/i18n.hh (hunk 4 above)
Comment (Baklap4) [reply game/i18n.hh:51]: Alright thanks :)

File: game/i18n.hh (hunk 5 above)
Comment (Baklap4) [reply game/i18n.hh:70]: Yuush will change

File: game/i18n.hh (hunk 6 above)
Comment (Baklap4) [reply game/i18n.hh:74]: It is set from the main where it reads the projectname shouldnt be null

File: game/screen.hh (hunk 7 above)
Comment (Baklap4) [reply game/screen.hh:109]: Alright will change

File: game/i18n.hh (hunk 2 above)
Comment (Baklap4) [reply game/i18n.hh:26]: Hmm for now i'll keep it in, can't really find a way to toggle between GETTEXT_FOUND or not which is what it should toggle

**Comment by Baklap4** [comment]
@Lord-Kamina could you gives this a try on macos?
Note; you'll have to proper install performous before you're able to check this.. since languages only get properly installed on install and not on build.. Maybe we need to fix that..

**Comment by Lord-Kamina** [comment]
I'll check it later, but keep in mind that in macOS, the app is always "installed", the installation folder is the app bundle. Please take a look at the launcher and the directory tree I'm attaching now so you can get an idea, then either you modify or tell how I need to modify it to use this.

[tree.txt](https://github.com/performous/performous/files/4447245/tree.txt)

**Comment by Lord-Kamina** [comment]
Nevermind, it works with the macOS bundle using the current structure.

However, switching to a language and then back to English doesn't work. It saves the setting but doesn't update the UI until you restart performous.

**Comment by Lord-Kamina** [comment]
That said, there's two changes I'd lie to see in this:

1) An "Automatic" or similar mode, which essentially keeps current behavior; in mac (and I believe perhaps in Windows also) it would automatically detect OS language and use that if available (User-set language from the UI should still win but I think reverting the English should be the last option)

And 2) This is purely cosmetic, but possibly you can use "virtOptName" I added to MenuOptions way back when I made the code to change audio backend, that way maybe we could use proper language names for displaying in the user UI?

**Comment by Tronic** [comment]
Btw, force-pushes make following the development harder. I would suggest not rewriting history / rebasing while on a PR, but only when merging to master.

**Comment by Baklap4** [comment]
> An "Automatic" or similar mode, which essentially keeps current behavior; in mac (and I believe perhaps in Windows also) it would automatically detect OS language and use that if available (User-set language from the UI should still win but I think reverting the English should be the last option)

If i'm correct it should get the system locale when nothing is within the `config.xml`
As for the English, that's a bit of a 'problem' since we actually don't have an English language. We have a Performous.pot file, which is basicly a po template file. Yes this is currently written in english but as far as the application is concerned it could be russian aswell. It's just a mapping between keys within the code and the languages files we have (which don't list english).

So i think we have to add a definition for the english language to get this fixed.
Agreed?

File: game/i18n.hh [hunk 12]
```diff
@@ -2,26 +2,74 @@
...
+			std::clog << "locale/notice: No given language.. using default system language: '" << lan << "''" << std::endl;
+			return lan;
+		}
+
+		auto allLanguages = GetAllLanguages();
+		auto foundLanguage = std::find(allLanguages.begin(), allLanguages.end(), language);
+		auto returnLang = std::string();
+
+		if (foundLanguage == allLanguages.end())
+			returnLang = "en_US.UTF-8";
```
Comment (Lord-Kamina) [inline game/i18n.hh:63]: I meant here

File: game/i18n.hh [hunk 13]
```diff
@@ -2,26 +2,74 @@
...
+	TranslationEngine(std::string package) : m_package(package) {
+		auto lang = boost::locale::util::get_system_locale(true);
+
+		std::clog << "locale/debug: Checking if language is set in config." << std::endl;
+		if(config["game/language"].i() >= 0) {
+			lang = config["game/language"].getEnumStringValueAtIndex(config["game/language"].i());
+			std::clog << "locale/debug: Language in configuration found: " << lang << std::endl;
+		}
+
+		if (lang.empty()) lang = "en_US.UTF-8";
```
Comment (Lord-Kamina) [inline game/i18n.hh:23]: And here

File: game/i18n.hh (hunk 12 above)
Comment (Baklap4) [reply game/i18n.hh:63]: I've added those since the system language doesn't have to be one of the languages we have translations for.

**Comment by Baklap4** [comment]
Correct as a first time there will be a `-1` i have no idea how to get rid of this.. :/

Regarding the userfriendly part of the names; i agree but i'm not quite sure how to do this with the current configuration system in place

**Comment by Lord-Kamina** [comment]
> Correct as a first time there will be a `-1` i have no idea how to get rid of this.. :/
> 
> Regarding the userfriendly part of the names; i agree but i'm not quite sure how to do this with the current configuration system in place

You could maybe use faux, dynamically generated values? Take a look at what I did in `populateBackends()`

**Comment by Baklap4** [comment]
> You could maybe use faux, dynamically generated values

@Lord-Kamina i've fixed that.

Also the it now auto selects the language as it used to do.

If your system uses anything we don't support it falls back to English.

It also enables translating the languages list into the translated language name.

Some things can maybe be cleaned up a bit but it's 3 am atm here so i just had to push it.

File: game/configuration.cc [hunk 14]
```diff
@@ -439,3 +494,17 @@ void populateBackends (const std::list<std::string>& backendList) {
...
+
+void populateLanguages(const std::map<std::string, std::string>& languages, bool refreshOutDated) {
+	ConfigItem& languageConfig = config["game/language"];
+	if (refreshOutDated) {
+		languageConfig.removeAllEnums();
+	}
+	for (auto const& language : languages) {
+		languageConfig.addEnum(language.second);
+	}
+	static std::string selectedLanguage = std::string();
```
Comment (OznOg) [inline game/configuration.cc:506]: why is that static? BTW the initialization is useless.

File: game/fs.cc [hunk 15]
```diff
@@ -12,6 +12,7 @@
 #include <mutex>
 #include <set>
 #include <sstream>
+#include <regex>
```
Comment (OznOg) [inline game/fs.cc:15]: is that realy needed in here?

File: game/game.hh [hunk 16]
```diff
@@ -75,6 +76,8 @@ class Game: public Singleton <Game> {
 	void drawLogo();
 	///global playlist access
 	PlayList& getCurrentPlayList() { return currentPlaylist; }
+	void setLanguage(const std::string language) { m_translationEngine.setLanguage(language, true); };
```
Comment (OznOg) [inline game/game.hh:79]:
```suggestion
	void setLanguage(const std::string &language) { m_translationEngine.setLanguage(language, true); };
```

File: game/game.hh [hunk 17]
```diff
@@ -75,6 +76,8 @@ class Game: public Singleton <Game> {
...
+	const std::string getCurrentLanguage() { return m_translationEngine.getCurrentLanguage().second; };
```
Comment (OznOg) [inline game/game.hh:80]:
```suggestion
	std::string getCurrentLanguage() { return m_translationEngine.getCurrentLanguage().second; };
```

File: game/game.hh [hunk 18]
```diff
@@ -107,6 +110,7 @@ private:
 	AnimValue m_dialogTimeOut;
 	// Dialog members
 	std::unique_ptr<Dialog> m_dialog;
+	TranslationEngine& m_translationEngine;
```
Comment (OznOg) [inline game/game.hh:113]: shoudl/could it be const?

File: game/i18n.hh [hunk 19]
```diff
@@ -2,26 +2,130 @@
...
-			std::locale::global(gen("en_US.UTF-8"));
-		}
+	TranslationEngine(const char *package) : m_package(package) {
+
+		initializeAllLanguages();
+		populateLanguages(GetAllLanguages());
+
+		std::clog << "locale/debug: Checking if language is set in config." << std::endl;
+
+		auto prefValue = config["game/language"].getEnumName();
```
Comment (OznOg) [inline game/i18n.hh:23]: why not storing english tong name instead of enums?

File: game/i18n.hh [hunk 20]
```diff
@@ -2,26 +2,130 @@
...
+		auto lang = getLanguageByHumanReadableName(prefValue);
+
+		setLanguage(lang);
 	};
 	static bool enabled() {
 		return true;
 	};
 
+	static ConfigItem& translationConfig() {
```
Comment (OznOg) [inline game/i18n.hh:32]: you are creating a singleton; is that really *needed* why not just returning the configIten by value?

File: game/i18n.hh [hunk 21]
```diff
@@ -2,26 +2,130 @@
...
+		static ConfigItem& backend = config["game/language"];
+		return backend;
+	}
+
+	void initializeAllLanguages() {
+		auto path = getLocaleDir().string();
```
Comment (OznOg) [inline game/i18n.hh:38]: shouldn't you keep path as is instead of converting?

File: game/i18n.hh [hunk 22]
```diff
@@ -2,26 +2,130 @@
...
+		else {
+			m_currentLanguage = { language , getLanguageByKey(language) };
+		}
+
+		try {
+			std::locale::global(m_gen(m_currentLanguage.first));
+			std::cout << "locale/notice: Current language is: '" << m_currentLanguage.second << "'" << std::endl;
+			populateLanguages(GetAllLanguages(true), true);
+		}
+		catch (...) {
```
Comment (OznOg) [inline game/i18n.hh:64]: is that really throwing *anything*?

File: game/i18n.hh [hunk 23]
```diff
@@ -2,26 +2,130 @@
...
+	}
+
+	std::string getLanguageByHumanReadableName(const std::string& language) {
+		if (language == _("Auto")) {
+			return boost::locale::util::get_system_locale(true);
+		}
+
+		auto allLanguages = GetAllLanguages();	
+		auto languageKey = std::string();
+		for (auto it = allLanguages.begin(); it != allLanguages.end(); ++it) {
```
Comment (OznOg) [inline game/i18n.hh:77]: use range for loop?

File: game/i18n.hh [hunk 24]
```diff
@@ -2,26 +2,130 @@
...
+		}
+
+		auto allLanguages = GetAllLanguages();
+		if (allLanguages.find(languageKey) == allLanguages.end()) {
+			return allLanguages["en_US.UTF-8"];
+		}
+		return allLanguages.at(languageKey);
+	}
+
+	const std::pair<std::string, std::string> getCurrentLanguage() { return m_currentLanguage; }
```
Comment (OznOg) [inline game/i18n.hh:99]: if you really want by value
```suggestion
	 std::pair<std::string, std::string> getCurrentLanguage() { return m_currentLanguage; }
```
or by reference:
```suggestion
	const std::pair<std::string, std::string> &getCurrentLanguage() { return m_currentLanguage; }
```

File: game/i18n.hh [hunk 25]
```diff
@@ -2,26 +2,130 @@
...
+
+	std::map<std::string, std::string> GetAllLanguages(bool refresh = false) {
+		if (refresh) {
+			m_languages.clear();
+		}
+		if (m_languages.size() != 0) return m_languages;
+		m_languages.emplace(std::pair("None", _("Auto")));
```
Comment (OznOg) [inline game/i18n.hh:106]: I think 
```suggestion
		m_languages = { {"None", _("Auto") }, ...}; 
```
should work

File: game/i18n.hh [hunk 26]
```diff
@@ -2,26 +2,130 @@
...
+		m_languages.emplace(std::pair("pl_PL.UTF-8", _("Polish")));
+		m_languages.emplace(std::pair("pt_PT.UTF-8", _("Portuguese")));
+		m_languages.emplace(std::pair("sv_SE.UTF-8", _("Swedish")));
+		m_languages.emplace(std::pair("zh_CN.UTF-8", _("Chinese")));
+		return m_languages;
+	}
+
+private:
+	std::pair<std::string, std::string> m_currentLanguage;
+	std::string m_package;
```
Comment (OznOg) [inline game/i18n.hh:128]: not really clear what all this stands for...

File: game/fs.cc (hunk 15 above)
Comment (Baklap4) [reply game/fs.cc:15]: Yup. Within `fs_paths.win.inc` we use a regex and we can't add the include in that file

File: game/game.hh (hunk 18 above)
Comment (Baklap4) [reply game/game.hh:113]: When making it const i'm getting:
```
Error (active)	E1086	the object has type qualifiers that are not compatible with the member function "TranslationEngine::setLanguage" C:\Users\Arjan\source\repos\performous\performous2\game\game.hh#79	
```

File: game/i18n.hh (hunk 21 above)
Comment (Baklap4) [reply game/i18n.hh:38]: Boost locale `generator.add_messages_path()` only supports std::string

File: game/i18n.hh (hunk 22 above)
Comment (Baklap4) [reply game/i18n.hh:64]: image
I guess we can catch runtime exception agreed?

File: game/i18n.hh (hunk 25 above)
Comment (Baklap4) [reply game/i18n.hh:106]: Using that notation now 👍

File: game/i18n.hh (hunk 26 above)
Comment (Baklap4) [reply game/i18n.hh:128]: it's the performous name. It's used for reference performous.mo files for every language and it comes from the main method

File: game/configuration.cc (hunk 14 above)
Comment (Baklap4) [reply game/configuration.cc:506]: Removed the initialization and static

File: game/fs.cc (hunk 15 above)
Comment (Baklap4) [reply game/fs.cc:15]: Yup, fs_paths.win.inc has a regex match and the include cannot be in there

File: game/game.hh (hunk 16 above)
Comment (Baklap4) [reply game/game.hh:79]: Changed to reference

File: game/game.hh (hunk 17 above)
Comment (Baklap4) [reply game/game.hh:80]: Removed the const

File: game/game.hh (hunk 18 above)
Comment (Baklap4) [reply game/game.hh:113]: Cannot be const due to `std::string getCurrentLanguage() { return m_translationEngine.getCurrentLanguage().second; }`

File: game/i18n.hh (hunk 19 above)
Comment (Baklap4) [reply game/i18n.hh:23]: Hmmm @Lord-Kamina suggested it this way like he did with Audio backends

File: game/i18n.hh (hunk 20 above)
Comment (Baklap4) [reply game/i18n.hh:32]: This method gets called from the config WriteConfig method. We don't have any reference to TranslationEngine there. Currently looking into it if it's necessairy

File: game/i18n.hh (hunk 22 above)
Comment (Baklap4) [reply game/i18n.hh:64]: Catching runtime exception now

File: game/i18n.hh (hunk 23 above)
Comment (Baklap4) [reply game/i18n.hh:77]: Fixed

File: game/i18n.hh (hunk 24 above)
Comment (Baklap4) [reply game/i18n.hh:99]: Did it by reference

File: game/i18n.hh (hunk 25 above)
Comment (Baklap4) [reply game/i18n.hh:106]: Yup works, nice cleanup

File: game/i18n.hh (hunk 26 above)
Comment (Baklap4) [reply game/i18n.hh:128]: It's the name of the application "performous" it's used as the *.mo files are all called Performous.mo once installed in different folders. It's given from the main method

File: game/i18n.hh (hunk 20 above)
Comment (Baklap4) [reply game/i18n.hh:32]: Removed it as it wasn't necessairy

File: game/game.hh (hunk 18 above)
Comment (OznOg) [reply game/game.hh:113]: Returns a copy THUS can be const

File: game/fs.cc (hunk 15 above)
Comment (OznOg) [reply game/fs.cc:15]: ok

File: game/game.hh [hunk 27]
```diff
@@ -75,6 +76,8 @@ class Game: public Singleton <Game> {
 	void drawLogo();
 	///global playlist access
 	PlayList& getCurrentPlayList() { return currentPlaylist; }
+	void setLanguage(const std::string& language) { m_translationEngine.setLanguage(language, true); };
+	std::string getCurrentLanguage() { return m_translationEngine.getCurrentLanguage().second; };
```
Comment (OznOg) [inline game/game.hh:80]:
```suggestion
	std::string getCurrentLanguage() const { return m_translationEngine.getCurrentLanguage().second; };
```

File: game/i18n.hh [hunk 28]
```diff
@@ -2,26 +2,126 @@
...
+		}
+
+		auto allLanguages = GetAllLanguages();
+		if (allLanguages.find(languageKey) == allLanguages.end()) {
+			return allLanguages["en_US.UTF-8"];
+		}
+		return allLanguages.at(languageKey);
+	}
+
+	const std::pair<std::string, std::string>& getCurrentLanguage() { return m_currentLanguage; }
```
Comment (OznOg) [inline game/i18n.hh:91]:
```suggestion
	const std::pair<std::string, std::string>& getCurrentLanguage() const { return m_currentLanguage; }
```

**Comment by ooshlablu** [comment]
I can confirm that this works. I can change the language and anything that has correct translation strings is converted appropriately after the settings are saved.

File: game/configuration.cc [hunk 29]
```diff
@@ -130,6 +132,13 @@ std::string const ConfigItem::getValue() const {
 		else std::clog << "audio/warning: Currently selected audio backend is unavailable on this system, will default to Auto." << std::endl;
 		return "Auto";
 	}
+	if (this->getName() == "game/language") {
+		int autoLanguageType = 1337;
+		static int val = std::get<int>(m_value);
```
Comment (OznOg) [inline game/configuration.cc:137]: why static? seem awkward here....

File: game/configuration.cc [hunk 30]
```diff
@@ -415,6 +449,27 @@ int PaHostApiNameToHostApiTypeId (const std::string& name) {
 	throw std::runtime_error("Invalid PortAudio HostApiTypeId Specified.");
 }
 
+unsigned int LanguageToLanguageId(const std::string& name) {
```
Comment (OznOg) [inline game/configuration.cc:452]: now we are in c++17, time to use std::optional<>

File: game/configuration.cc [hunk 31]
```diff
@@ -415,6 +449,27 @@ int PaHostApiNameToHostApiTypeId (const std::string& name) {
...
+	if (name == _("Asturian")) return 1;
+	if (name == _("Danish")) return 2;
+	if (name == _("German")) return 3;
+	if (name == _("English")) return 4;
+	if (name == _("Spanish")) return 5;
```
Comment (OznOg) [inline game/configuration.cc:457]: what are all those magic values refering to (1, 2, 3, ...)? is that some order somewhere?

File: game/configuration.hh [hunk 32]
```diff
@@ -43,6 +43,7 @@ class ConfigItem {
 	std::string const& getShortDesc() const { return m_shortDesc; } ///< get the short description for this ConfigItem
 	std::string const& getLongDesc() const { return m_longDesc; } ///< get the long description for this ConfigItem
 	void addEnum(std::string name); ///< Dynamically adds an enum to all values
+	void removeAllEnums(); /// Removes all the enum values from this entry.
```
Comment (OznOg) [inline game/configuration.hh:46]: I'm affraid the naming "enume values" is quite missleading, there is no enum in there, isn't there?

Comment (Baklap4) [reply game/configuration.hh:46]: So to get back to this. Within our `schema.xml` we define "enums" these "enums" are basicly a label with an integer backing it up. All of the enums defined in there aren't actual Enum class values throughout our C++ codebase.

The name enum for this is poorly choosen or the usage is poorly applied. As for now i think we'll have to live with this. But i can feel a configuration.cc/hh overhaul coming.

File: game/configuration.cc (hunk 31 above)
Comment (Baklap4) [reply game/configuration.cc:456]: As stated below the "enum" parent needs an integer value to work correctly. The numbers here are arbitrary and doesn't refer to anything except the selected language in the parent scope of the schema.xml

File: game/configuration.cc (hunk 30 above)
Comment (Baklap4) [reply game/configuration.cc:451]: Not quite sure what you want me to do here. All the cases are handled in this method

File: game/configuration.cc (hunk 29 above)
Comment (Baklap4) [reply game/configuration.cc]: Removed the static and the check on the next line as it didn't provide any value.

File: game/configuration.cc (hunk 30 above)
Comment (OznOg) [reply game/configuration.cc:451]: don't mind, I think this can be done in a future rework

File: game/configuration.cc (hunk 31 above)
Comment (OznOg) [reply game/configuration.cc:456]: ok

File: game/configuration.hh (hunk 32 above)
Comment (OznOg) [reply game/configuration.hh:46]: ok

File: game/configuration.cc (hunk 29 above)
Comment (OznOg) [reply game/configuration.cc]: ok

**Review by OznOg — APPROVED** [review]
LGTM

**Pull request merged**